SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp

```diff
@@ -2173,3 +2258,189 @@ void USpatialReceiver::PeriodicallyProcessIncomingRPCs()
 		}
 	}, GetDefault<USpatialGDKSettings>()->QueuedIncomingRPCWaitTime, true);
 }
+
+bool USpatialReceiver::NeedToLoadClass(const FString& ClassPath)
+{
+	return FindObject<UClass>(nullptr, *ClassPath, false) == nullptr;
+}
+
+FString USpatialReceiver::GetPackagePath(const FString& ClassPath)
+{
+	return FSoftObjectPath(ClassPath).GetLongPackageName();
+}
+
+void USpatialReceiver::OnAsyncPackageLoaded(const FName& PackageName, UPackage* Package, EAsyncLoadingResult::Type Result)
+{
+	FString PackageNameStr = PackageName.ToString();
+	if (!AsyncLoadingPackages.Contains(PackageNameStr))
+	{
+		UE_LOG(LogSpatialReceiver, Error, TEXT("USpatialReceiver::OnAsyncPackageLoaded: Package loaded but no entry in AsyncLoadingPackages. Package: %s"), *PackageNameStr);
+		return;
+	}
+
+	TArray<Worker_EntityId> Entities = AsyncLoadingPackages.FindAndRemoveChecked(PackageNameStr);
+
+	for (Worker_EntityId Entity : Entities)
+	{
+		if (IsEntityWaitingForAsyncLoad(Entity))
+		{
+			UE_LOG(LogSpatialReceiver, Log, TEXT("Finished async loading package %s for entity %lld."), *PackageNameStr, Entity);
+
+			CriticalSectionSaveState CriticalSectionState;
+			if (bInCriticalSection)
+			{
+				SaveCriticalSection(CriticalSectionState);
+			}
+			bool bInCriticalSectionSaved = bInCriticalSection;
+			bInCriticalSection = true;
+
+			EntityWaitingForAsyncLoad AsyncLoadEntity = EntitiesWaitingForAsyncLoad.FindAndRemoveChecked(Entity);
+			PendingAddEntities.Add(Entity);
+			PendingAddComponents = MoveTemp(AsyncLoadEntity.InitialPendingAddComponents);
+			LeaveCriticalSection();
+
+			for (QueuedOpForAsyncLoad& Op : AsyncLoadEntity.PendingOps)
+			{
+				HandleQueuedOpForAsyncLoad(Op);
+			}
+
+			bInCriticalSection = bInCriticalSectionSaved;
+			if (bInCriticalSection)
+			{
+				RestoreCriticalSection(CriticalSectionState);
+			}
+		}
+	}
+}
+
+bool USpatialReceiver::IsEntityWaitingForAsyncLoad(Worker_EntityId Entity)
+{
+	return EntitiesWaitingForAsyncLoad.Contains(Entity);
+}
+
+void USpatialReceiver::QueueAddComponentOpForAsyncLoad(const Worker_AddComponentOp& Op)
+{
+	EntityWaitingForAsyncLoad& AsyncLoadEntity = EntitiesWaitingForAsyncLoad.FindChecked(Op.entity_id);
+
+	QueuedOpForAsyncLoad NewOp = {};
+	NewOp.AcquiredData = Worker_AcquireComponentData(&Op.data);
```

Comment:
Just wondering why we have AcquiredData as a member on QueuedOpForAsyncLoad. It seems we duplicate the data here NewOp.Op.op.add_component.data. I see we do Worker_ReleaseComponentData later on but perhaps we can do it here and avoid holding onto the pointer?

Reply:
We need to hold onto the pointer because that is the mechanism that C API provides to extend the lifetime of the schema objects, and we need to release the pointer after we're done.